### DISCUSSION TITLE
Increased DNS Requests After Enabling Istio Egress Gateways

### DISCUSSION THREAD

**Discussion by volkorny** [istio/istio#51455] [Q&A] [answered]
Hello!

We have the `ndots:1` configuration enabled for our applications. Before migrating to Istio, our DNS resolutions of external domains were done with a single request to CoreDNS.

After migrating to Istio, everything worked fine for months. Three months ago, we decided to enable [Egress Gateways](https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway/) to control egress traffic and set the `outboundTrafficPolicy:mode:REGISTRY_ONLY` mode. We noticed that the number of requests to CoreDNS increased by 4-5 times. In the CoreDNS logs, we see a series of requests instead of a single request:

```
  [INFO] 10.x.x.55:39786 - 26360 "A IN taskrouter.twilio.com.istio-ingress.svc.cluster.local. udp 71 false 512" NXDOMAIN qr,aa,rd 164 0.000116158s
  [INFO] 10.x.x.55:39786 - 7696 "A IN taskrouter.twilio.com.svc.cluster.local. udp 57 false 512" NXDOMAIN qr,aa,rd 150 0.000092738s
  [INFO] 10.x.x.55:39786 - 62817 "A IN taskrouter.twilio.com.cluster.local. udp 53 false 512" NXDOMAIN qr,aa,rd 146 0.0000795s
  [INFO] 10.x.x.55:39786 - 3610 "A IN taskrouter.twilio.com.us-west-2.compute.internal. udp 66 false 512" NXDOMAIN qr,rd,ra 179 0.000313721s
  [INFO] 10.x.x.55:39786 - 62146 "A IN taskrouter.twilio.com. udp 39 false 512" NOERROR qr,aa,rd,ra 397 0.000054419s
```
Our resolv.conf file still has `ndots:1`, but it seems like this option doesn't work with Istio.
```
/app $ cat /etc/resolv.conf
search staging-mesh.svc.cluster.local svc.cluster.local cluster.local us-west-2.compute.internal
nameserver 172.20.0.10
options ndots:1
/app $
```
The increased load on CoreDNS is causing timeout errors in our environment.

Please advise if it is possible to decrease the number of requests to CoreDNS to the level it was before enabling the [Egress Gateways](https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway/).
Thanks!

**Comment by husnialhamdani** [comment]
Maybe you can enable Istio Smart DNS proxy in Istio https://istio.io/latest/blog/2020/dns-proxy/#reduced-load-on-your-dns-servers-w-faster-resolution, https://istio.io/latest/docs/ops/configuration/traffic-management/dns-proxy/.

**Reply by volkorny** [reply]
My configuration contains this and dns proxy looks like enabled:
```
meshConfig:
  defaultConfig:
    proxyMetadata:
      # Enable basic DNS proxying
      ISTIO_META_DNS_CAPTURE: "true"
      # Enable automatic address allocation, optional
      ISTIO_META_DNS_AUTO_ALLOCATE: "true"
```

**Comment by howardjohn** [comment, marked as answer]
I cannot reproduce. Are you sure the queries this is coming from has ndots:1? Note it could be the ServiceEntry you created is applying to every proxy in the cluster (or at least, many proxies), and not all have ndots=1 set. You can find who the `10.x.x.55:39786` IP is in the query and double check?

BTW, the DNS proxy does not help as you found; that only applies to your own application, not DNS requests send from Istio itself.

**Reply by volkorny** [reply]
@howardjohn Thanks for your hint!
I have checked the IPs and PODs and found that only the ingress PODs have ndots:5, causing them to make 5 requests to CoreDNS. I changed this setting to ndots:1, which resulted in nearly a 25% decrease in the load on CoreDNS.

However, I also discovered that every Istio-proxy instance (sidecar) periodically (approximately every minute) makes CoreDNS requests to resolve every ServiceEntry domain we have in our mesh, even if the service in the POD never makes requests to that domain. Why does Envoy do this so frequently? Is it possible to increase this interval to 5-10 minutes?

**Reply by howardjohn** [reply]
https://github.com/envoyproxy/envoy/issues/20562 has more context on why it does. Currently we always hardcode to respect DNS TTL: https://github.com/istio/istio/blob/3006870b5f795b7cfa9ac6509a1cb83e93153853/pilot/pkg/networking/core/cluster_builder.go#L312-L313. Note DnsRefreshRate, which you can configure, only applies for failed queries due to that.

**Reply by volkorny** [reply]
Thanks for sharing this info. It helps us better understand how Istio works.

It appears that "respect DNS TTL" isn't functioning correctly in our mesh. I set a domain with a TTL of 3600, but the sidecars are still making requests to CoreDNS every 1-2 minutes for this domain. Could you provide some guidance on how to investigate this issue?
We use Istio: 1.20.2

**Reply by howardjohn** [reply]
Hmm, if you set debug or trace logs in envoy I believe it logs how long it is waiting

**Reply by volkorny** [reply]
I see messagess "DNS refresh rate reset for DOMAIN, refresh rate NUMBER ms" where NUMBER is different values from 1000 to 30000 for different DOMANs

**Reply by volkorny** [reply]
[file-domain.txt](https://github.com/user-attachments/files/16053926/file-domain.txt)
File from single POD grep'ed by SOMEDOMAIN
From 2024-07-01T13:46:05 to 2024-07-01T13:46:58

**Reply by howardjohn** [reply]
have you tried manually querying the DNS from the pod (with dig or kdig, should be in the proxy image)?That will show the ttl. Maybe it really is 30s

**Reply by volkorny** [reply]
I have created new DNS record with TTL 7200 secnds.
But I still have on logs:
DNS refresh rate reset for DOMAIN, refresh rate XXX (XXX are: 30000 ms, 9000 ms, 29000 ms, 24000 ms)
but never more than 30000 ms

**Reply by volkorny** [reply]
In CoreDNS logs I see every pod in our meah make requests to this 7200-seconds record 1-2 timers per minute...

**Reply by howardjohn** [reply]
the reason I asked to do a query with dig is it could be the ttl is modified by kube-dns

**Reply by volkorny** [reply]
yes, it really modified by kube-dns, thanks
